File: problems/73.set-matrix-zeroes/solution.ts
function setZeroes(matrix: number[][]): number[][] {
  const zeroX: Set<number> = new Set();
  const zeroY: Set<number> = new Set();

  matrix.forEach((row, i) =>
    row.forEach((num, j) => {
      if (num === 0) {
        zeroX.add(i);
        zeroY.add(j);
      }
    }),
  );

  zeroX.forEach((x: number) => (matrix[x] = Array(matrix[0].length).fill(0)));

  zeroY.forEach((y) => matrix.forEach((m) => (m[y] = 0)));

  return matrix;
}

export { setZeroes };

File: problems/73.set-matrix-zeroes/solution.go
package main

func setZeroes(matrix [][]int) [][]int {
	firstColHasZero := false

	for i := range matrix {
		if matrix[i][0] == 0 {
			firstColHasZero = true
		}

		for j := 1; j <= len(matrix[0])-1; j++ {
			if matrix[i][j] == 0 {
				matrix[i][0] = 0
				matrix[0][j] = 0
			}
		}
	}

	for i := 1; i < len(matrix); i++ {
		for j := 1; j < len(matrix[0]); j++ {
			if matrix[i][0] == 0 || matrix[0][j] == 0 {
				matrix[i][j] = 0
			}
		}
	}

	if matrix[0][0] == 0 {
		for k := range matrix[0] {
			matrix[0][k] = 0
		}
	}

	if firstColHasZero {
		for k := range matrix {
			matrix[k][0] = 0
		}
	}

	return matrix
}
